### PR TITLE
Kupccgsd identify parameter

### PR DESCRIPTION
**Pull request by dchung0741**
Identify particular parameters in kupgccsd ansatz

- (Fix): Non-Singlet excitation kUpGCCSD ansatz: parameter count reduced to $3N(N-1)/2$ where $N$ is the number of spatial orbitals.
- (Add): Singlet excitation kUpGCCSD ansatz: parameter count reduced to $N(N-1)$ where $N$ is the number of spatial orbitals.

### CODE REVIEW

**Comment by kwkbtr** [comment]
Can you write some comments in the PR description if there are breaking changes? (e.g. number/order of parameters etc.)

(By the way, #181 is already merged but could you add comments about breaking changes there too?)

**Comment by dchung0741** [comment]
I just added comments on this PR page and the PR page for the UCCSD

File: packages/openfermion/quri_parts/openfermion/ansatz/kupccgsd.py [hunk 1]
```diff
@@ -55,53 +60,142 @@ def __init__(
         fermion_qubit_mapping: OpenFermionQubitMapping = jordan_wigner,
         trotter_number: int = 1,
         delta_sz: int = 0,
+        singlet_excitation: bool = False,
```
Comment (kwkbtr) [inline packages/openfermion/quri_parts/openfermion/ansatz/kupccgsd.py:68]: Can you add docstring for `singlet_excitation`?

Comment (dchung0741) [reply packages/openfermion/quri_parts/openfermion/ansatz/kupccgsd.py:68]: I just added the docstring for it

File: packages/openfermion/quri_parts/openfermion/ansatz/kupccgsd.py [hunk 2]
```diff
@@ -50,6 +50,9 @@ class KUpCCGSD(ImmutableLinearMappedUnboundParametricQuantumCircuit):
           :class:`Operator`
         trotter_number: Number for first-order Trotter product formula.
         delta_sz: Changes of spin in the excitation.
+        singlet_excitation: If ``True``, single excitations that start
+            from and end at the same spatial orbitals will share the same
+            circuit parameter.
```
Comment (kwkbtr) [inline packages/openfermion/quri_parts/openfermion/ansatz/kupccgsd.py]: I'm a bit confused: in my understanding "singlet excitation" and "single excitations" are different. Is this description correct?

Comment (dchung0741) [reply packages/openfermion/quri_parts/openfermion/ansatz/kupccgsd.py]: In the context of kUpGCCSD this description is correct. Singlet excitation means that the excitation operator $T$ should satisfy $[T, S_x] = [T, S_y]= [T, S_z] = 0$. For UCCSD, this condition imposes constraint on both single and double excitation. For kUpGCCSD, the double excitation operator is paired, so it is singlet automatically. Thus the singlet condition only imposes constraint on the single excitation part in the way I describe docstring.

Sorry if the names are a bit confusing, should I describe it in detail in the docstring?

Comment (kwkbtr) [reply packages/openfermion/quri_parts/openfermion/ansatz/kupccgsd.py]: I see, thanks! Mentioning the following point in the docstring may be helpful for readers:

> For kUpGCCSD, the double excitation operator is paired, so it is singlet automatically

Comment (dchung0741) [reply packages/openfermion/quri_parts/openfermion/ansatz/kupccgsd.py]: I just added the comment to the docstring.

**Review by kwkbtr — APPROVED** [review]

**Pull request merged**